### PR TITLE
feat/utm analytics

### PR DESCRIPTION
**Pull request by blaggacao**
##### ATTENTION: Requires at least https://github.com/frappe/frappe/commit/7abd0b15b643c73e4f58afb3a92206b194118779 together with https://github.com/frappe/frappe/pull/27801/commits/5c12091385d8f91db4e50735caf5a9addba7ef40 (check if these commits are in your deployed frappe branch before migrating!)

 # Add UTM parameters to transactional doctypes

 This pull request adds UTM (Urchin Tracking Module) parameters to all erp transactional doctypes to better track marketing efforts and their impact on the buisness cycle.

 ## Changes

 - Added `utm_campaign`, `utm_source`, `utm_medium`, and `utm_content` fields to all tx doctypes.
 - Updated all indirect references that I could find (either by failing tests or by 3 months running in production)

 ## Details

 The following UTM fields have been added:

 - `utm_campaign`: Link field to track specific campaigns
 - `utm_source`: Link field to track the source of the opportunity
 - `utm_medium`: Link field to track the marketing medium
 - `utm_content`: Data field to track specific content within a campaign

 These fields replace the previous `campaign` and `source` fields, providing more detailed tracking capabilities for marketing efforts.

 ## Notes

 This change may require updates to any custom scripts or reports that rely on the previous `campaign` or `source` fields. Users should update their custom code accordingly.

CRM ("messaging") Campaigns will be henceforth automatically linked to UTM ("tracking") Campaigns, and so does the migration. Here is how it looks like from now on:
 
 
image

## Further Changes

- https://github.com/frappe/ecommerce_integrations should consider to use the [new upstream data utils](https://github.com/frappe/frappe/pull/26891/files#diff-593e371f8b132db49a3762060b70e1a32493b58ae5673086132ecba513938c74) tools to sync UTM data from order and provide a better Ecomm integration.
- Our yet to be released ECWID integration makes proper use of this, already.

`no-docs`

### CODE REVIEW

File: erpnext/stock/doctype/delivery_note/delivery_note.py [hunk 1]
```diff
@@ -121,15 +120,13 @@ class DeliveryNote(SellingController):
...
 		shipping_rule: DF.Link | None
-		source: DF.Link | None
 		status: DF.Literal["", "Draft", "To Bill", "Completed", "Return Issued", "Cancelled", "Closed"]
 		tax_category: DF.Link | None
 		tax_id: DF.Data | None
 		taxes: DF.Table[SalesTaxesandCharges]
 		taxes_and_charges: DF.Link | None
 		tc_name: DF.Link | None
 		terms: DF.TextEditor | None
-		territory: DF.Link | None
```
Comment (blaggacao) [inline erpnext/stock/doctype/delivery_note/delivery_note.py:132]: this is a cherry-pick error

File: erpnext/accounts/doctype/pos_invoice/pos_invoice.json [hunk 2]
```diff
@@ -1558,12 +1542,41 @@
...
    "fieldtype": "Check",
    "label": "Update Billed Amount in Delivery Note"
+  },
+  {
+   "fieldname": "column_break_gpiw",
+   "fieldtype": "Column Break"
+  },
+  {
+   "fieldname": "utm_medium",
+   "print_hide": 1,
```
Comment (blaggacao) [inline erpnext/accounts/doctype/pos_invoice/pos_invoice.json:1552]: note how all new fields have print_hide set (except on lead)

**Comment by blaggacao** [comment]
> Fears are nothing more than a state of mind.   - Napoleon Hill 

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by blaggacao** [comment]
> Watch your thoughts, they become your words Watch your words, they become your actions Watch your actions, they become your habits Watch your habits, they become your character Watch your character, it becomes your destiny.  - 

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by blaggacao** [comment]
> You may say Im a dreamer, but Im not the only one, I hope someday you will join us, and the world will live as one.  - John Lennon

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by blaggacao** [comment]
> I have just three things to teach: simplicity, patience, compassion. These three are your greatest treasures.   - Lao Tzu 

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by ruthra-kumar** [comment]
@blaggacao 

Maybe provide more details on what this feature is.

**Comment by blaggacao** [comment]
@ruthra-kumar, this depends on the merge of [PR #26891](https://github.com/frappe/frappe/pull/26891).

This implementation ensures consistent end-to-end marketing tracking via UTM (Urchin Tracking Module). For instance, if an inbound UTM specifies `?utm_campaign=foo&utm_source=fb&utm_medium=display`, all subsequent business documents can be analyzed based on this UTM. This allows grouping Sales Orders by source (e.g., `fb` for "Facebook"), campaign, medium, and content.

This consistency enables us to measure marketing effectiveness across the **entire** sales lifecycle accurately. As mentioned, the prerequisite data structures for this traceability are implemented in [PR #26891](https://github.com/frappe/frappe/pull/26891).

**Comment by blaggacao** [comment]
> The important thing is this: to be able at any moment to sacrifice what we are for what we could become.  - Charles Dubois

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by blaggacao** [comment]
> Obstacles are those frightful things you see when you take your eyes off your goal.  - Henry Ford

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by devdiogenes** [comment]
Hello. Do you know why your pull request not runned the checks? The pull requests I made in the ERPNext is with the same issue

**Comment by blaggacao** [comment]
I don't know. But this PR needs an upstream merge anyways, first. So I'm not worried about tests, atm. Iirc, they were all passing.

**Comment by blaggacao** [comment]
> Our distrust is very expensive.  - Ralph Emerson

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by devdiogenes** [comment]
> I don't know. But this PR needs an upstream merge anyways, first. So I'm not worried about tests, atm. Iirc, they were all passing.

Thanks for answering

**Comment by blaggacao** [comment]
> An invincible determination can accomplish almost anything and in this lies the great distinction between great men and little men.  - Thomas Fuller 

---

<sub>Kindly help move this PR forward. Many thanks from Yours Sincerely!</sub>

**Comment by blaggacao** [comment]
I'm putting this on auto-merge cause the upstream change is likely to break master, due to a (now) stale doctype reference. I've run this in production for 3 months now and it works as expected increasing the range of marketing (automation) capabilities for our business.

I'll stand by attentively should anything come up with this change.

**Comment by mudux** [comment]
Just run migrate and got this:

`"#### query\nSELECT `name` FROM `tabPatch Log` WHERE `patch`='erpnext.patches.v15_0.migrate_to_utm_analytics' AND `skipped`=0 ORDER BY `modified` DESC LIMIT 1\n####"
'#### query\ncommit\n####'
'#### query\nSTART TRANSACTION\n####'
Executing erpnext.patches.v15_0.migrate_to_utm_analytics in demo.ruphasoft.com (demo)

        Remove Lead Source doctype and use UTM Source Instead
        Ensure that for each Campaign, a UTM Campaign is also set

'#### query\nSTART TRANSACTION\n####'
'#### query\nSELECT `source_name`,`details` FROM `tabLead Source`\n####'
'#### query\nrollback\n####'
'#### query\nSTART TRANSACTION\n####'
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 114, in <module>
    main()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 20, in main
    click.Group(commands=commands)(prog_name="bench")
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 29, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/apps/frappe/frappe/commands/site.py", line 651, in run_patch
    frappe.modules.patch_handler.run_single(module, force=force or context.force)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 152, in run_single
    return execute_patch(patchmodule, method, methodargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 188, in execute_patch
    _patch()
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/patches/v15_0/migrate_to_utm_analytics.py", line 27, in execute
    insert_query.run()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/query_builder/utils.py", line 87, in execute_query
    result = frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 234, in sql
    self._cursor.execute(query, values)
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/cursors.py", line 153, in execute
    result = self._query(query)
             ^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/cursors.py", line 322, in _query
    conn.query(q)
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/connections.py", line 563, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/connections.py", line 825, in _read_query_result
    result.read()
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/connections.py", line 1199, in read
    first_packet = self.connection._read_packet()
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/connections.py", line 775, in _read_packet
    packet.raise_for_error()
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/protocol.py", line 219, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/home/frappe/frappe-bench/env/lib/python3.11/site-packages/pymysql/err.py", line 150, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table 'demo.tabUTM Source' doesn't exist")`

**Comment by blaggacao** [comment]
@mudux acked, on it. Thanks for the report! Is there a way I can reach out on telegram if need be?

**Comment by blaggacao** [comment]
Can you execute these steps:
- Update frappe at the same time? (`UTM Source` is a **new** frappe website doctype)
- If that does not work:
  - Update frappe first?
  - Update erpnext second?

**Comment by mudux** [comment]
i think my problem is erpnext is for some reason on develop, while frappe on v15..I'll fix that

**Comment by blaggacao** [comment]
@mudux please quickly confirm for clarity in posterity: where you able to solve the issue by updating frappe at the same time? Thx

**Comment by mudux** [comment]
@blaggacao it was an error on my part to be on develop branch of erpnext, i switched back and the issue resolved but I havent yet tested develop branch.

**Comment by blaggacao** [comment]
You'd also want to include https://github.com/frappe/frappe/pull/27801/commits/5c12091385d8f91db4e50735caf5a9addba7ef40 as of now, thanks to the report of a fellow frappe user.

**Pull request merged**